### DISCUSSION TITLE
What model loaders work with GGUF models other than llama.cpp?

### DISCUSSION THREAD

**Discussion by CalculonPrime** [oobabooga/text-generation-webui#4066] [Q&A]
I selected a GGUF model because it's the new format, but most loaders in the dropdown don't seem to like it.  Will any of them work besides the default of llama.cpp? [Edit: I see that the only other loader that works is ctransformers, but it burns up my CPUs compared with llama.cpp for not any better performance.]

Was it a mistake to choose a GGUF model or is it just going to be slow no matter what with a 70B model? For 13900K (32 cores), 128GB main memory, and GTX 4060 (8GB), I'm getting maybe one word per second when it responds.  It takes some time to start responding.  I have "low-vram" and mul_mat_q options checked in llama.cpp, and all other settings at default.

Also, how can I view my current tokens/second? [Edit: OK I see in the shell that it reports I'm getting < 1 token/second.  How can I improve this?  Is there a way to run a 70b GGUF model in 4-bit mode with llama.cpp?]

**Comment by viperwasp** [comment]
For starters GGUF 70B have a whole bunch of different quantizations. Running them in 4-bit would be mean like downloading 4_S or 4_M would be the models you need to download. Yes I think only llama.cpp and ctransformers work. 

I suggest you set threads to the proper number based on your CPU. I know for a 12700K 8 threads should be optimal. Anyone can correct me if I am wrong. I think you need to set threads to be equal to your primary cores. So you have 32 cores but not all of those are primary so change your threads. 

Wait? oh no? it's two threads per core? Now I'm messed up? I will look it up later. Having a brain fart. But the only other thing I believe you can do is offload onto your GPU. A 4060 8GB can't offload much for 70B. It will only speed it up a little. But the layers option you need to increase that and offload as much of the AI into VRAM as possible. Without stability issues. It just depends on what else your doing like multitasking. I generally offload onto the GPU until my VRAM use is about 90-96% when running the model. So give it a test prompt. Hit generate. If your VRAM is about 90-96% that should be fine. You want to give it some overhead. If you run into issues lower your layers and try again. This will increase your speed by a little. Besides this I don't have any other suggestions etc. 

The fastest speed you could likely get away with and still have good results would be 4_S and offloading onto GPU with Ctransformers or llama.cpp. I generally use Ctransformers now. Somestimes a fresh reinstall of Ooba can help too. I ran into an issues that a fresh install speed things up a tiny bit.
